### PR TITLE
Auton drive distance

### PR DESCRIPTION
**Pull request by MrAlpaca**

### CODE REVIEW

File: Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java [hunk 1]
```diff
@@ -46,96 +45,79 @@ public void setPIDSourceType(PIDSourceType pidSource)
...
 
-		pidLeft = new PIDController(0, 0, 0, new PIDSource()
-		{
+				SmartDashboard.putNumber("Current Distance", currentDist);
 
-			public void setPIDSourceType(PIDSourceType pidSource)
-			{
-				return;
-			}
+				if (currentDist > dist + (double) config
```
Comment (MrAlpaca) [inline Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java]: Oh god, fix this if clause

Comment (MrAlpaca) [inline Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java]: Use the pid's method onTarget

File: Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java [hunk 2]
```diff
@@ -46,96 +45,79 @@ public void setPIDSourceType(PIDSourceType pidSource)
...
-	protected void execute()
-	{
 		currentTime = Timer.getFPGATimestamp() - initTime;
-		
-		SmartDashboard.putNumber("Motion profile velocity",
-				motion.getVelocity(currentTime));
+		currentDist = 0;
 
-		set();
+		pid.enable();
```
Comment (MrAlpaca) [inline Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java:90]: Why do we need to re-enable the controller every iteration?

File: Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java [hunk 3]
```diff
@@ -46,96 +45,79 @@ public void setPIDSourceType(PIDSourceType pidSource)
...
-	protected abstract void set();
+	protected void execute()
+	{
+	}
 
-	protected abstract boolean isFinished();
+	protected boolean isFinished()
+	{
+		double measuredSpeed = (Robot.chassis.getLeftSpeed() + Robot.chassis.getRightSpeed()) / 2;
+		return measuredSpeed == 0 && currentDist > dist/2;
```
Comment (MrAlpaca) [inline Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java]: Why is it okay to finish after moving only half of the distance?

File: Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java [hunk 4]
```diff
@@ -46,96 +45,79 @@ public void setPIDSourceType(PIDSourceType pidSource)
...
 		}, new PIDOutput()
 		{
 
 			public void pidWrite(double output)
 			{
-				pidLeftOutput = output;
+				currentTime = Timer.getFPGATimestamp() - initTime;
+				profileVelocity = motion.getVelocity(currentTime);
+
+				SmartDashboard.putNumber("Motion Planner Velocity",
```
Comment (karkason) [inline Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java]: Remove this or move it to SDB.

File: Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java (hunk 3 above)
Comment (karkason) [inline Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java]: In order to not finish as soon as we start.
Change this to pid.isOnTarget() .

File: Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java [hunk 5]
```diff
@@ -46,96 +45,79 @@ public void setPIDSourceType(PIDSourceType pidSource)
...
+				return PIDSourceType.kDisplacement;
 			}
 		}, new PIDOutput()
 		{
 
 			public void pidWrite(double output)
 			{
-				pidLeftOutput = output;
+				currentTime = Timer.getFPGATimestamp() - initTime;
+				profileVelocity = motion.getVelocity(currentTime);
```
Comment (karkason) [inline Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java]: Why are profileVelocity and currentTime global variables ?

File: Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java (hunk 1 above)
Comment (karkason) [inline Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java]: onTarget didn't work yesterday, need figure out way.

File: Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java [hunk 6]
```diff
@@ -46,96 +45,79 @@ public void setPIDSourceType(PIDSourceType pidSource)
...
-		pidLeft.setOutputRange(-1, 1);
+		pid.setOutputRange(-1, 1);
 
-		// PID values are the same for all subclasses for until we finish
-		// testings
-		pidRight.setPID((double) config.get("chassis_PIDRight_KP") / 1000,
-				(double) config.get("chassis_PIDRight_KI") / 1000,
-				(double) config.get("chassis_PIDRight_KF") / 1000,
-				(double) config.get("chassis_PIDRight_KF") / 1000);
+		pid.setAbsoluteTolerance(
```
Comment (karkason) [inline Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java:78]: Maybe use setPercentTolerance to fix the onTarget error?

File: Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java [hunk 7]
```diff
@@ -46,96 +45,79 @@ public void setPIDSourceType(PIDSourceType pidSource)
...
-			public void setPIDSourceType(PIDSourceType pidSource)
-			{
-				return;
-			}
+				if (currentDist > dist + (double) config
+						.get("chassis_DriveDistance_PID_Tolerance")
+						&& currentDist < dist - (double) config
+								.get("chassis_DriveDistance_PID_Tolerance"))
+				{
+					pid.reset();
```
Comment (MrAlpaca) [inline Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java]: I heard that we're not using an integral term anyway. Why then do we need to reset the pid controller?

Comment (karkason) [inline Robot Code/src/org/usfirst/frc/team3316/robot/commands/chassis/autonomous/DriveDistance.java]: AKA remove this

**Pull request merged**